src/methods/widen.js

```diff
@@ -0,0 +1,309 @@
+import { Generator } from "../utils/generator.js";
+import { AbstractValue, ObjectValue, Value } from "../values/index.js";
+
+import invariant from "../invariant.js";
+
+export class WidenImplementation {
+  _widenArrays(
+    realm: Realm,
+    v1: void | Array<Value> | Array<{ $Key: void | Value, $Value: void | Value }>,
+    v2: void | Array<Value> | Array<{ $Key: void | Value, $Value: void | Value }>
+  ): Array<Value> | Array<{ $Key: void | Value, $Value: void | Value }> {
+    let e = (v1 && v1[0]) || (v2 && v2[0]);
+    if (e instanceof Value) return this._widenArraysOfValues(realm, (v1: any), (v2: any));
+    else return this._widenArrayOfsMapEntries(realm, (v1: any), (v2: any));
+  }
+
+  _widenArrayOfsMapEntries(
+    realm: Realm,
+    a1: void | Array<{ $Key: void | Value, $Value: void | Value }>,
+    a2: void | Array<{ $Key: void | Value, $Value: void | Value }>
+  ): Array<{ $Key: void | Value, $Value: void | Value }> {
+    let empty = realm.intrinsics.empty;
+    let n = Math.max((a1 && a1.length) || 0, (a2 && a2.length) || 0);
+    let result: Array<{ $Key: void | Value, $Value: void | Value }> = [];
+    for (let i = 0; i < n; i++) {
+      let { $Key: key1, $Value: val1 } = (a1 && a1[i]) || { $Key: empty, $Value: empty };
+      let { $Key: key2, $Value: val2 } = (a2 && a2[i]) || { $Key: empty, $Value: empty };
+      if (key1 === undefined && key2 === undefined) {
+        result[i] = { $Key: undefined, $Value: undefined };
+      } else {
+        let key3 = this.widenValues(realm, key1, key2);
+        invariant(key3 instanceof Value);
+        let val3 = this.widenValues(realm, val1, val2);
+        invariant(val3 === undefined || val3 instanceof Value);
+        result[i] = { $Key: key3, $Value: val3 };
+      }
+    }
+    return result;
+  }
+
+  _widenArraysOfValues(realm: Realm, a1: void | Array<Value>, a2: void | Array<Value>): Array<Value> {
+    let n = Math.max((a1 && a1.length) || 0, (a2 && a2.length) || 0);
+    let result = [];
+    for (let i = 0; i < n; i++) {
+      let wv = this.widenValues(realm, (a1 && a1[i]) || undefined, (a2 && a2[i]) || undefined);
+      invariant(wv === undefined || wv instanceof Value);
+      result[i] = wv;
+    }
+    return result;
+  }
+
+  // Returns a new effects summary that includes both e1 and e2.
+  widenEffects(realm: Realm, e1: Effects, e2: Effects): Effects {
+    let [result1, , bindings1, properties1, createdObj1] = e1;
+    let [result2, , bindings2, properties2, createdObj2] = e2;
+
+    let result = this.widenResults(realm, result1, result2);
+    let bindings = this.widenBindings(realm, bindings1, bindings2);
+    let properties = this.widenPropertyBindings(realm, properties1, properties2, createdObj1, createdObj2);
+    let createdObjects = new Set();
+    let generator = new Generator(realm); // code subject to widening will be generated somewhere else
+    return [result, generator, bindings, properties, createdObjects];
```

Comment:
Is `createdObjects` always the empty set after widening. Naively I would expect it to be the union.

Reply:
I scratched my head about it for a while. A union does not seem quite right to me since it will not converge onto a fixpoint. It is a bit surprising to think of the empty set as Top, but since createdObjects is used only during joins and for a simplistic form of escape analysis, the empty set actually has the right semantics. I'll add a comment.

Reply:
So the empty set implies "no information" which in the context of an escape analysis means everything might escape?